src/bt.rs

```diff
@@ -0,0 +1,358 @@
+        for bit in path {
+            if let Some(next_node) = node {
+                node = if !bit {
+                    &mut next_node.left
+                } else {
+                    &mut next_node.right
+                };
+            } else {
+                return Err(BinaryTreeError::UnreachablePath(value));
+            }
+        }
+
+        // Checks whether the node already has a value,
+        // If so, it replaces with the new value, and returns the old one.
+        // otherwise, node is available for insertion, and `None` is returned.
+        if let Some(current) = node {
+            Ok(Some(mem::replace(&mut current.value, value)))
+        } else {
+            *node = Some(Box::new(Node::new(value)));
+            Ok(None)
+        }
+    }
+
+    /// Gets a reference to the value located at the end of the path.
+    /// [`None`] is returned if no values are found during the traversal of the tree
+    /// along the path.
+    pub fn get(&self, path: &Path) -> Option<&V> {
+        self.root.as_deref().and_then(|mut node| {
+            for bit in path {
+                match if !bit { &node.left } else { &node.right } {
+                    None => return None,
+                    Some(next_node) => node = next_node,
+                };
+            }
+            Some(&node.value)
+        })
+    }
+}
+
+impl<V> Default for BinaryTree<V> {
+    fn default() -> Self {
+        Self {
+            root: Option::default(),
+        }
+    }
+}
+
+impl<V: Encode> Encode for BinaryTree<V> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        let mut stack = Vec::new();
+        stack.push(&self.root);
+
+        // Nodes are stored following a pre-order traversal.
+        while let Some(elem) = stack.pop() {
+            match elem {
+                None => CodecMarker::None.encode(bytes)?,
+                Some(node) => {
+                    CodecMarker::Node.encode(bytes)?;
+                    node.value.encode(bytes)?;
+                    stack.push(&node.right);
+                    stack.push(&node.left);
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<V: Decode> Decode for BinaryTree<V> {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let mut root = None;
+
+        let mut stack = Vec::new();
+        stack.push(&mut root);
+
+        // Decode nodes in a pre-order traversal.
+        while let Some(elem) = stack.pop() {
+            match CodecMarker::decode(bytes)? {
+                CodecMarker::None => (),
+                CodecMarker::Node => {
+                    *elem = Some(Box::new(Node::new(V::decode(bytes)?)));
+                    let node = elem.as_mut().unwrap();
+                    stack.push(&mut node.right);
+                    stack.push(&mut node.left);
+                }
+            };
+        }
+
+        Ok(Self { root })
+    }
+}
+
+impl<V: Display> Display for BinaryTree<V> {
```

Comment:
Holy crap this is pretty!! However the output will contain secret shares when used with VIDPF. We are treating secret shared data as something you don't want to inadvertently log. Thus implementation of `Debug` or `Display` should either not contain any secret shares or should be fenced behind the "test-util" feature such that it's less likely to be compiled in production.

We should also decide whether we're OK with special characters or would rather stick to ascii. I'm agnostic, but @tgeoghegan may have thoughts here for example.

Reply:
cc/ @divergentdave: Should `Debug` implementations have unicode characters in them?

Reply:
I have changed the unicode symbols for ascii chars.
Coincidentally, printing, for example a `BinaryTree<u32>`, produces a valid YAML output.